dali/pipeline/data/tensor_vector.h

```diff
@@ -0,0 +1,182 @@
+ * providing an uniform way of handling a collection/batch of tensors_.
+ *
+ * Propagates Buffer calls to every tensor uniformly
+ *
+ * @tparam Backend
+ */
+template <typename Backend>
+class TensorVector {
+ public:
+  TensorVector() {}
+  explicit TensorVector(int batch_size) : tensors_(batch_size, nullptr) {
+    for (auto &t : tensors_) {
+      t = std::make_shared<Tensor<Backend>>();
+    }
+  }
+
+  TensorVector(const TensorVector &) = delete;
+  TensorVector &operator=(const TensorVector &) = delete;
+
+  auto operator[](size_t pos) {
+    return tensors_[pos];
+  }
+
+  auto operator[](size_t pos) const {
+    return tensors_[pos];
+  }
+
+  auto begin() noexcept {
+    return tensors_.begin();
+  }
+
+  auto begin() const noexcept {
+    return tensors_.begin();
+  }
+
+  auto cbegin() const noexcept {
+    return tensors_.cbegin();
+  }
+
+  auto end() noexcept {
+    return tensors_.end();
+  }
+
+  auto end() const noexcept {
+    return tensors_.end();
+  }
+
+  auto cend() const noexcept {
+    return tensors_.cend();
+  }
+
+  auto size() const noexcept {
+    return tensors_.size();
+  }
+
+  kernels::TensorListShape<> shape() const {
+    if (tensors_.empty()) {
+      return {};
+    }
+    kernels::TensorListShape<> result(tensors_.size(), tensors_[0]->ndim());
+    for (size_t i = 0; i < tensors_.size(); i++) {
+      result.set_tensor_shape(i, tensors_[i]->shape());
+    }
+    return result;
+  }
+
+  DLL_PUBLIC inline void Resize(const kernels::TensorListShape<> &new_shape) {
+    // N.B. There probably is nothing wrong with adjusting batchsize for give TensorVector
+    // (sparse tensor list), but the semantics of what type and pinned status should
+    // the new elements have or having some of them allocated and the new ones not
+    // complicates the logic even further so we disallow it
+    DALI_ENFORCE(tensors_.empty() || static_cast<int>(tensors_.size()) == new_shape.size(),
+                 "Changing the batch size is prohibited. It should be set once.");
+    if (tensors_.empty()) {
+      allocate_tensors(new_shape.size());
+    }
+    for (int i = 0; i < new_shape.size(); i++) {
+      tensors_[i]->Resize(new_shape[i]);
+    }
+  }
+
+  inline void set_type(const TypeInfo &new_type) {
+    type_ = new_type;
+    for (auto t : tensors_) {
+      t->set_type(new_type);
+    }
+  }
+
+  inline TypeInfo type() const {
+    // TODO(klecki): do we enforce the same type between elements, or prohibit mixing APIs?
+    return type_;
+  }
+
+  inline void set_pinned(const bool pinned) {
+    // Store the value, in case we pin empty vector and later call Resize
+    pinned_ = pinned;
+    for (auto &t : tensors_) {
+      t->set_pinned(pinned);
+    }
+  }
+
+  inline bool is_pinned() const {
+    // TODO(klecki): do we enforce the same pin status between elements, or prohibit mixing APIs?
+    bool pinned = true;
+    for (auto &t : tensors_) {
```

Comment:
I don't think it's actually meaningful if tensors' `is_pinned()` returns mixed results - and I don't know if it's correct to assume that a vector with 99 pinned and 1 non-pinned tensor is non-pinned.

Reply:
I guess I can return a value based on the first one. If the user decides to have inconsistent state, that's his problem.